File: ts/src/fallbackStream.ts
/*******************************************************************************
 * Copyright (C) 2019 - 2023 DiffusionData Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/

import { connect, datatypes, topics, Session } from 'diffusion';

// example showcasing how to add a fallback stream that receives updates for topics that
// don't have a value stream
export async function fallbackStreamExample(): Promise<void> {

    // Data Types are exposed from the top level Diffusion namespace. It is often easier
    // to assign these directly to a local variable.
    const stringDataType = datatypes.string();
    const TopicSpecification = topics.TopicSpecification;
    const TopicType = topics.TopicType;

    // Connect to the server. Change these options to suit your own environment.
    // Node.js does not accept self-signed certificates by default. If you have
    // one of these, set the environment variable NODE_TLS_REJECT_UNAUTHORIZED=0
    // before running this example.
    const session: Session = await connect({
        host: 'diffusion.example.com',
        port: 443,
        secure: true,
        principal: 'control',
        credentials: 'password'
    });

    try {
        // 1. Create a String topic type
        await session.topics.add('topic/foo', new TopicSpecification(TopicType.STRING));
        // Once the topic is added successfully then update it with a new value
        await session.topicUpdate.set('topic/foo', stringDataType, "foo-string");
    } catch (error) {
        console.log("Fail to update topic 'foo': ", error);
    }

    try {
        // Create another String topic type
        await session.topics.add('topic/bar', new TopicSpecification(TopicType.STRING));
        await session.topicUpdate.set('topic/bar', stringDataType, 'bar-string');
    } catch (error) {
        console.log("Fail to update topic 'bar': ", error);
    }

    try {
        // And another String topic type
        await session.topics.add('topic/baz', new TopicSpecification(TopicType.STRING));
        await session.topicUpdate.set('topic/baz', stringDataType, "baz-string");
    } catch (error) {
        console.log("Fail to update topic 'baz': ", error);
    }

    // 2. Register a value stream for receiving String value update
    session.addStream('topic/foo', stringDataType)
        .on('value', function(topic, specification, newValue, oldValue) {
            console.log("Received update ", newValue);
        });


    // 3. Register a fallback value stream to receive update for any topics that don't have a stream register
    // In this example, the falback stream will receive updates from 'topic/bar' and 'topic/baz'
    session.addFallbackStream(stringDataType)
        .on('value', function(topic, specification, newValue, oldValue) {
            console.log("Received update ", newValue);
        });

    // 4. Subscribe
    session.select('?topic//');
}

File: c/examples/pub_sub/subscribe/fallback_stream.c
/**
 * Copyright © 2024 DiffusionData Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This example is written in C99. Please use an appropriate C99 capable compiler
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "diffusion.h"
#include "utils.h"
MUTEX_DEF

static int on_subscription(
    const char *const topic_path,
    const TOPIC_SPECIFICATION_T *const specification,
    void *context)
{
    printf("Subscribed to %s.\n", topic_path);
    return HANDLER_SUCCESS;
}

static int on_unsubscription(
    const char *const topic_path,
    const TOPIC_SPECIFICATION_T *const specification,
    NOTIFY_UNSUBSCRIPTION_REASON_T reason,
    void *context)
{
    printf("Unsubscribed from %s.\n", topic_path);
    return HANDLER_SUCCESS;
}

static int on_value(
    const char *const topic_path,
    const TOPIC_SPECIFICATION_T *const specification,
    DIFFUSION_DATATYPE datatype,
    const DIFFUSION_VALUE_T *const old_value,
    const DIFFUSION_VALUE_T *const new_value,
    void *context)
{
    char *old_value_json_string;
    if (old_value == NULL) {
        old_value_json_string = "NULL";
    }
    else {
        to_diffusion_json_string(old_value, &old_value_json_string, NULL);
    }

    char *new_value_json_string;
    if (new_value == NULL) {
        new_value_json_string = "NULL";
    }
    else {
        to_diffusion_json_string(new_value, &new_value_json_string, NULL);
    }

    printf(
        "%s changed from %s to %s.\n",
        topic_path, old_value_json_string, new_value_json_string
    );

    if (old_value != NULL) {
        free(old_value_json_string);
    }

    if (new_value != NULL) {
        free(new_value_json_string);
    }
    return HANDLER_SUCCESS;
}

static int on_subscribe(
    SESSION_T*session,
    void *context)
{
    printf("Subscription request received and approved by the server.\n");
    MUTEX_BROADCAST
    return HANDLER_SUCCESS;
}


void run_example(
    const char *url,
    const char *principal,
    CREDENTIALS_T *credentials)
{
    MUTEX_INIT

    SESSION_T *session = session_create(
        url, principal, credentials, NULL, NULL, NULL
    );

    utils_create_json_topic(
        session,
        "my/topic/path",
        "{\"diffusion\": [ \"data\", \"some data\" ] }"
    );

    utils_create_json_topic(
        session,
        "my/other/topic/path",
        "{\"diffusion\": [ \"data\", \"more data\" ] }"
    );

    VALUE_STREAM_T value_stream = {
        .datatype = DATATYPE_JSON,
        .on_subscription = on_subscription,
        .on_unsubscription = on_unsubscription,
        .on_value = on_value
    };

    add_fallback_stream(session, &value_stream);

    const char *topic_selector = "?my//";

    SUBSCRIPTION_PARAMS_T params = {
            .topic_selector = topic_selector,
            .on_subscribe = on_subscribe
    };

    subscribe(session, params);
    MUTEX_WAIT

    // Sleep for a bit to see the notifications
    sleep(2);

    utils_create_json_topic(
        session,
        "my/additional/topic/path",
        "{\"diffusion\": [ \"data\", \"even more data\" ] }"
    );

    utils_create_json_topic(
        session,
        "this/topic/path/will/not/be/picked/up",
        "{\"diffusion\": [ \"no data\" ] }"
    );

    // Sleep for a bit to see the notifications
    sleep(2);


    session_close(session, NULL);
    session_free(session);
    MUTEX_TERMINATE
}